clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -2184,9 +2178,139 @@ void handleFailure(Throwable throwable) {
+                    .collect(Collectors.toMap(topicRequest -> topicRequest.name(), topicRequest -> topicRequest, (t1, t2) -> t1, TreeMap::new));
+
+            TopicDescription partiallyFinishedTopicDescription = null;
+            TopicDescription nextTopicDescription = null;
+
+            @Override
+            DescribeTopicPartitionsRequest.Builder createRequest(int timeoutMs) {
+                DescribeTopicPartitionsRequestData request = new DescribeTopicPartitionsRequestData()
+                    .setTopics(new ArrayList<>(pendingTopics.values()))
+                    .setResponsePartitionLimit(options.partitionSizeLimitPerResponse());
+                if (partiallyFinishedTopicDescription != null) {
+                    request.setCursor(new DescribeTopicPartitionsRequestData.Cursor()
+                        .setTopicName(partiallyFinishedTopicDescription.name())
+                        .setPartitionIndex(partiallyFinishedTopicDescription.partitions().size())
+                    );
+                }
+                return new DescribeTopicPartitionsRequest.Builder(request);
+            }
+
+            @Override
+            void handleResponse(AbstractResponse abstractResponse) {
+                DescribeTopicPartitionsResponse response = (DescribeTopicPartitionsResponse) abstractResponse;
+                DescribeTopicPartitionsResponseData.Cursor responseCursor = response.data().nextCursor();
+
+                for (DescribeTopicPartitionsResponseTopic topic : response.data().topics()) {
+                    String topicName = topic.name();
+                    Errors error = Errors.forCode(topic.errorCode());
+
+                    KafkaFutureImpl<TopicDescription> future = topicFutures.get(topicName);
+
+                    if (error != Errors.NONE) {
+                        future.completeExceptionally(error.exception());
+                        pendingTopics.remove(topicName);
+                        if (responseCursor != null && responseCursor.topicName().equals(topicName)) {
+                            responseCursor = null;
+                        }
+                        continue;
+                    }
+
+                    TopicDescription currentTopicDescription = getTopicDescriptionFromDescribeTopicsResponseTopic(topic, nodes);
+
+                    if (partiallyFinishedTopicDescription != null && partiallyFinishedTopicDescription.name().equals(topicName)) {
+                        partiallyFinishedTopicDescription.partitions().addAll(currentTopicDescription.partitions());
+                        continue;
+                    }
+
+                    if (responseCursor != null && responseCursor.topicName().equals(topicName)) {
+                        // This is the first time this topic being pointed by the cursor.
+                        nextTopicDescription = currentTopicDescription;
+                        continue;
+                    }
```

Comment:
why not just 
```
  if (partiallyFinishedTopicDescription != null) {
    if (partiallyFinishedTopicDescription.name().equals(topicName)) {
      // We've got unfinished description from previous invocation.
      partiallyFinishedTopicDescription.partitions().addAll(currentTopicDescription.partitions());
      continue;
    }
  } else {
    if (responseCursor != null && responseCursor.topicName().equals(topicName)) {
      // This topic is part of the next cursor, so it may not have all partitions.
      partiallyFinishedTopicDescription = currentTopicDescription;
      continue;
    }
  }

```

then do we need the nextTopicDescription?

Reply:
The problem is that in the same batch of results, we can have both some partitions for a topic partially finished and for a new topic pointed by the cursor. If we want to close the future for the partially finished after the loop, we have to store the cursor topic partitions somewhere(nextTopicDescription) during the loop.
However, it may be easier if we can close the future during the loop because the API guarantees that the partially finished will be the first in the result, and the cursor topic shows at the end.